packages/vitest/src/runtime/mocker.ts

```diff
@@ -335,7 +335,38 @@ export class VitestMocker {
     return this.request(dep)
   }
 
+  private wrapFactoryInProxy(id: string, importer: string, factory?: () => unknown) {
+    if (!factory)
+      return factory
+
+    const exportHandler = {
+      get(target: Record<string, any>, prop: any) {
+        const val = target[prop]
+
+        // 'then' can exist on non-Promise objects, need nested instanceof check for logic to work
+        if (prop === 'then') {
+          if (target instanceof Promise)
+            return target.then.bind(target)
+        }
+        else if (val === undefined) {
+          throw new Error(`[vitest] No "${prop}" export is defined on the "${id}" mock defined here: ${importer}`)
+        }
+
+        return val
+      },
+    }
+
+    const factoryHandler = {
+      apply(target: () => Object) {
+        return new Proxy(target(), exportHandler)
+      },
+    }
+
+    return new Proxy(factory, factoryHandler) as () => unknown
+  }
+
   public queueMock(id: string, importer: string, factory?: () => unknown) {
+    factory = this.wrapFactoryInProxy(id, importer, factory)
```

Comment:
Can't we just wrap exports here:

https://github.com/jereklas/vitest/blob/9d5234ef74e092a902314d2baac7cfbdd6cb2729/packages/vitest/src/runtime/mocker.ts#L109

?

Reply:
And while we are here, let's throw an error, if factory returns non-object. (With a helper that it should return `{ default }`, if user wants to mock default export)

Reply:
I split the proxy objects. The export object is now happening where you suggested, and the factory proxy object is now returning an error if the return value is not an object.